### PR TITLE
[SYCL-MLIR][MathOps] Add `sycl::half` support

### PR DESCRIPTION
**Pull request by jopperm**
**Lowering:**
```mlir
%res = sycl.math.sqrt(%arg) : !sycl_half
```
⬇️
```mlir
%0 = sycl.mlir.unwrap(%arg) : !sycl_half to f16
%1 = math.sqrt(%0) : f16
%res = sycl.mlir.wrap(%1) : f16 to !sycl_half
```
⬇️
```llvm
%0 = extractvalue %"class.sycl::_V1::half" %arg, 0
%1 = call half @llvm.sqrt.f16(half %0)
%res = insertvalue %"class.sycl::_V1::half" undef, half %1, 0
```

### CODE REVIEW

File: mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp [hunk 1]
```diff
@@ -33,11 +33,35 @@ namespace {
...
+    if (type.isF32() || type.isF64()) {
+      rewriter.replaceOpWithNewOp<MathOpT>(op, adaptor.getOperands());
+      return success();
+    }
+
+    // `op` is using a SYCL-specific type; we need to unwrap the operands and
+    // wrap the result.
+    assert(isSYCLType(type));
+
+    if (type.isa<HalfType>()) {
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp:53]:
```suggestion
    if (isa<HalfType>(type)) {
```

File: mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp [hunk 2]
```diff
@@ -290,6 +290,7 @@ bool SYCLUnwrapOp::areCastCompatible(TypeRange inputs, TypeRange outputs) {
   Type sourceType = inputs.front();
   Type resultType = outputs.front();
   Type bodyType = getBodyType(sourceType);
+
```
Comment (victor-eds) [inline mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp:293]:
```suggestion
```

**Review by victor-eds — COMMENTED** [review]
Would it make sense to have a new conversion pattern to keep the original one clean?

File: mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp (hunk 2 above)
Comment (jopperm) [reply mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp:293]: Yeah, I forgot to add the newline in #10139 (`SYCLWrapOp::areCastCompatible` has it) -- can I maybe sneak that change in here, please? 😇

**Comment by jopperm** [comment]
> Would it make sense to have a new conversion pattern to keep the original one clean?

We could, but the downside would be bloating of the `patterns.add<...>` call (and again in the future for vectors and marrays). I think I'd rather keep the pattern per lowering-approach (now we only have 1:1, but for example there will be xyz-multipled-by-pi lowerings in the future), and handle the types inside. Maybe the type handling can be factored out with some clever template magic.

File: mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp [hunk 3]
```diff
@@ -33,11 +33,35 @@ namespace {
 template <typename SYCLOpT, typename MathOpT>
 struct OneToOneMappingPattern : public OpConversionPattern<SYCLOpT> {
   using OpConversionPattern<SYCLOpT>::OpConversionPattern;
+
   LogicalResult
   matchAndRewrite(SYCLOpT op, typename SYCLOpT::Adaptor adaptor,
```
Comment (sommerlukas) [inline mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp:38]: I agree with @victor-eds, having a separate pattern for non-MLIR-native types might make sense, especially given the name of this pattern (with wrap and unwrap, it's not really 1:1 anymore).

File: mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp (hunk 2 above)
Comment (victor-eds) [reply mlir-sycl/lib/Dialect/SYCL/IR/SYCLOps.cpp:293]: Allowed :smile:

File: mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp (hunk 3 above)
Comment (jopperm) [reply mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp:38]: Done.

**Review by sommerlukas — APPROVED** [review]

File: mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp
Comment (whitneywhtsang) [inline mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp:1]: are we missing the test case for wrap or unwrap lowering to llvm?

Comment (jopperm) [reply mlir-sycl/lib/Conversion/SYCLToLLVM/DPCPP.cpp:1]: Ah yes, currently these lowerings are tested only implicitly by the `math_funcs.cpp` tests. I'll add standalone tests.

File: mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp [hunk 4]
```diff
@@ -33,36 +33,75 @@ namespace {
...
+    if (isa<HalfType>(type)) {
+      auto loc = op.getLoc();
+      SmallVector<Value, 3> unwrappedOperands;
+      for (auto operand : adaptor.getOperands())
+        unwrappedOperands.push_back(
+            rewriter.create<SYCLUnwrapOp>(loc, rewriter.getF16Type(), operand));
+      auto math = rewriter.create<MathOpT>(loc, unwrappedOperands);
+      rewriter.replaceOpWithNewOp<SYCLWrapOp>(op, type, math.getResult());
+      return success();
+    }
```
Comment (victor-eds) [inline mlir-sycl/lib/Conversion/SYCLToMath/SYCLToMath.cpp:71]: Can we have separate `match` and `rewrite` for this and the above?

**Review by victor-eds — COMMENTED** [review]
:+1: to @whitneywhtsang's comment

**Review by victor-eds — APPROVED** [review]

**Review by whitneywhtsang — APPROVED** [review]

**Pull request merged**